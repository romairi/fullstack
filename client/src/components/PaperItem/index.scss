@import "../../scss/variables";

.paper_item_container {
  margin-bottom: 50px;
  margin-top: 50px;

  &:last-child {
    margin-bottom: 0;
  }

  .paper_item_card {
    padding: 20px 45px;
    transition: background-color 0.25s, color 0.25s, box-shadow 0.25s;
    box-shadow: 0 8px 42px -8px rgba(0, 0, 0, 0.2);

    &:hover {
      box-shadow: 0 8px 42px -8px rgba(0, 0, 0, 0.4);
    }

    .paper_item_title {
      font-family: 'Baloo 2', cursive;
    }

    .paper_item_summary {
      text-align: left;
      font-family: 'Tajawal', sans-serif;
    }

    .paper_item_published_authors {
      display: block;
      margin-bottom: 15px;
      font-family: 'Dancing Script', cursive;

      .authors_title {
        color: $primary_color;
      }
    }

    .paper_item_footer {
      display: flex;
      justify-content: space-between;
      font-family: 'Martel', serif;


      @media(max-width: $max-sm) {
        flex-direction: column-reverse;
      }

      .paper_item_footer_btn {
        font-family: 'Dancing Script', cursive;
      }

      .paper_item_meta {
        display: flex;
        flex-direction: column;
        align-items: start;
        @media(max-width: $max-sm) {
          margin-bottom: 7px;
        }

        .data_title {
          color: $primary_color;
        }
      }

    }

    .paper_item_save {
      display: flex;
      justify-content: flex-end;

      .btn_save {
        margin-top: 22px;
        font-family: 'Dancing Script', cursive;

        @media(max-width: $max-sm) {
          width: 100%;
        }
      }
    }
  }
}



