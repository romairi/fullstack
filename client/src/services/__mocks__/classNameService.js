module.exports = {
    buildClassName: jest.fn(() => 'COMPUTED_CLASS_NAME'),
};
