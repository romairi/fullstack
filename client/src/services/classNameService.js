export const buildClassName = classes => classes.filter(c => !!c).join(' ');
