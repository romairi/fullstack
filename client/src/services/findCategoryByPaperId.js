export function findCategoryByPaperId(categories, paperId) {
    return categories.find(c => c.paperItems.find(paper => paper.paperId === paperId));
}
