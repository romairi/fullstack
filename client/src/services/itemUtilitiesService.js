export function searchByFields(items, field) {
    return items.filter(item => [item.title, item.summary, ...item.authors.filter(a => Array.isArray(a) && a.length > 0).map(a => a[0])]
        .map(text => text.toLowerCase())
        .filter(text => text.includes(field.toLowerCase()))
        .length > 0
    );
}
