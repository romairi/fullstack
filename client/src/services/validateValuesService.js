export function reformatErrors(details) {
    return details.reduce((acc, cur) => {
        acc[cur.context.key] = cur.message;
        return acc;
    }, {});
}

export function validateValues(values, schema) {
    const results = schema.validate(values, {abortEarly: false});
    let errors = {};

    if (results.error) {
        errors = reformatErrors(results.error.details);
    }
    return errors;
}

