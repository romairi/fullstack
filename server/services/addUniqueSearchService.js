function formatSearchTags(listTags) {
    return listTags.map(tag => tag.toLowerCase()).sort().join();
}

function uniqueSearch(userSearches, includeTags, excludeTags) {
    const includeList = [];
    const excludeList = [];
    let i = 0;

    while (userSearches.length > i) {
        includeList.push(formatSearchTags(userSearches[i].include_tags));
        excludeList.push(formatSearchTags(userSearches[i].exclude_tags));
        i++;
    }

    const searchIncTags = formatSearchTags(includeTags.map(item => item));
    const searchExcTags = formatSearchTags(excludeTags.map(item => item));
    const foundInclude = includeList.find(item => item === searchIncTags);
    const foundExclude = excludeList.find(item => item === searchExcTags);
    const includeTagExists = searchIncTags.includes(foundInclude);
    const excludeTagExists = searchExcTags.includes(foundExclude);

    return (includeTagExists && excludeTagExists);
}

module.exports = {uniqueSearch};
