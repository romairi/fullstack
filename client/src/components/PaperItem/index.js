import React from 'react';
import './index.scss';
import {buildClassName} from "../../services/classNameService";
import Card from '@material-ui/core/Card';
import Button from "@material-ui/core/Button";
import SaveIcon from '@material-ui/icons/Save';
import DeleteIcon from '@material-ui/icons/Delete';


export function ActionButton({id, startIcon, children, onClick}) {
    return (<Button
        className="btn_save"
        variant="contained"
        color="primary"
        size="large"
        startIcon={startIcon}
        onClick={() => onClick(id)}
    >
        {children}
    </Button>);
}

function PaperItem(props) {
    const {
        className, title, summary, authors, pdfLink, publishedDate,
        updatedDate, onSaveButtonClicked, onRemoveButtonClicked, id, paperExist,
    } = props;

    const authorsNames = authors !== undefined ? authors.join(', ') : '';
    const buttonIcon = paperExist ? <DeleteIcon/> : <SaveIcon/>;
    const buttonOnClick = paperExist ? onRemoveButtonClicked : onSaveButtonClicked;
    const buttonText = paperExist ? 'Remove' : 'Save';

    return (
        <div className={buildClassName(['paper_item_container', className])}>
            <Card className="paper_item_card">
                <div className='paper_item_header'>
                    <h3 className='paper_item_title'>{title}</h3>
                </div>
                <p className='paper_item_summary'>
                    {summary}
                </p>
                <span className='paper_item_published_authors'>
                    <strong className="authors_title">Authors:</strong> {authorsNames}
                </span>
                <div className='paper_item_footer'>
                    <Button
                        className="paper_item_footer_btn"
                        component="a"
                        target='_blank'
                        href={pdfLink}
                        onClick={e => e.preventDefault}
                        color="secondary"
                        variant="contained"
                        rel="noreferrer">
                        PDF link
                    </Button>
                    <div className="paper_item_meta">
                        <span className='paper_item_published_date'>
                            <strong className="data_title">Published Date:</strong> {publishedDate}
                        </span>
                        <span className='paper_item_updated_date'>
                           <strong className="data_title">Updated Date:</strong> {updatedDate}
                        </span>
                    </div>
                </div>
                <div className="paper_item_save">
                    <ActionButton id={id} startIcon={buttonIcon} onClick={buttonOnClick}>
                        {buttonText}
                    </ActionButton>
                </div>
            </Card>
        </div>
    );
}

export default PaperItem;
