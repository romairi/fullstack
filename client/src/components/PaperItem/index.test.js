import React from 'react';
import renderer from 'react-test-renderer';
import PaperItem, {ActionButton} from './index';
import {buildClassName} from "../../services/classNameService";

jest.mock('../../services/classNameService');
// jest.mock('@material-ui/core/Card', () => () => <div className="Card"/>);
jest.mock('@material-ui/core/Button', () => () => <div className="Button"/>);
jest.mock('@material-ui/icons/Save', () => () => <div className="SaveIcon"/>);
jest.mock('@material-ui/icons/Delete', () => () => <div className="DeleteIcon"/>);

describe('PaperItem', () => {
    describe('render', () => {
        it('should render the props', () => {
            const component = renderer.create(
                <PaperItem/>
            );
            expect(component.toJSON()).toMatchSnapshot();
            expect(buildClassName.mock.calls.pop()).toMatchSnapshot();
        });
    });

    describe('ActionButton', () => {
        describe('render', () => {
            it('should render the children', () => {

                const component = renderer.create(
                    <ActionButton>
                        <div>CHILDREN</div>
                    </ActionButton>
                );

                expect(component.toJSON()).toMatchSnapshot();
            });
        })
    })
});

