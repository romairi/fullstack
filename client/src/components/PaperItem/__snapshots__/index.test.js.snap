// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaperItem ActionButton render should render the children 1`] = `
<div
  className="Button"
/>
`;

exports[`PaperItem render should render the props 1`] = `
<div
  className="COMPUTED_CLASS_NAME"
>
  <div
    className="MuiPaper-root MuiCard-root paper_item_card MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="paper_item_header"
    >
      <h3
        className="paper_item_title"
      />
    </div>
    <p
      className="paper_item_summary"
    />
    <span
      className="paper_item_published_authors"
    >
      <strong
        className="authors_title"
      >
        Authors:
      </strong>
       
      
    </span>
    <div
      className="paper_item_footer"
    >
      <div
        className="Button"
      />
      <div
        className="paper_item_meta"
      >
        <span
          className="paper_item_published_date"
        >
          <strong
            className="data_title"
          >
            Published Date:
          </strong>
           
        </span>
        <span
          className="paper_item_updated_date"
        >
          <strong
            className="data_title"
          >
            Updated Date:
          </strong>
           
        </span>
      </div>
    </div>
    <div
      className="paper_item_save"
    >
      <div
        className="Button"
      />
    </div>
  </div>
</div>
`;

exports[`PaperItem render should render the props 2`] = `
Array [
  Array [
    "paper_item_container",
    undefined,
  ],
]
`;
